ORMap and friends are delta-CRDTs (#22350)

gosubpl opened the pull request:
* ORMap
* LWWMap
* PNCounterMap
* ORMultiMap
* using deltas of value type in ORMap deltas whenever that value type supports deltas

gosubpl committed:
ORMap and friends have deltas (#22350)

gosubpl commented:
Refs #22350

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,49 @@ object ORMap {
+  }
+
+  // DeltaGroup is effectively a causally ordered list of individual deltas
+  final case class DeltaGroup[A, B <: ReplicatedData](ops: immutable.IndexedSeq[DeltaOp]) extends DeltaOp {
```
why didn't I make this `InternalApi` in ORSet?

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -100,8 +151,17 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         "`ORMap.put` must not be used to replace an existing `ORSet` " +
           "value, because important history can be lost when replacing the `ORSet` and " +
           "undesired effects of merging will occur. Use `ORMultiMap` or `ORMap.updated` instead.")
-    else
```
I think we can extend the above limitation to also cover other things that `RequiresCausalDeliveryOfDeltas` if that simplifies anything here

gosubpl replied:
I don't think this is particular to the *causal CRTDs*. All are affected, because `ORSet` indexing `ORMap` is causal, so replacements of full values are always causing anomalies. However, the effect of this is probably most visible when the value in question is itself an `ORSet`. On the other hand working with `updated` is a bit harder, as you need to supply the transformation function, which is helped by pre-defined maps - the _friends_ of `ORMap`: `ORMultiMap`, `PNCounterMap` and `LWWMap`. So I would probably leave that as is or, for consistency, either extend it to all values or to all friends. WDYT?

gosubpl replied:
I would keep this comment as is, because, actually, only `ORMultiMapWithDeltaDelta` is free from the anomalies here

patriknw replied:
ok

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -100,8 +151,17 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        case Some(d) ⇒
+          d.merge(PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → value), node))
+        case None ⇒
+          PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → value), node)
```
`PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → value), node)` is the same for both cases so can be defined in a `val` before the `match`

gosubpl replied:
fixed

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -125,11 +185,42 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+      (newValue, PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue), node))
+
+    val (newValue, modifiedDelta) =
+      if (oldValue.isInstanceOf[DeltaReplicatedData]) {
```
use `match` instead

gosubpl replied:
fixed - here and in another place

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +274,120 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
+    val removedKeysIterator = (tombstoneVals.keys ++ that.tombstoneVals.keys).toIterator
```
just `.iterator`

gosubpl replied:
thanks, fixed

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +274,120 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+      }
+    }
+    // tombstone the values for removed keys
+    new ORMap(mergedKeys, mergedValues, removedValues -- mergedKeys.elementsMap.keys.toIterator)
```
`keysIterator`

gosubpl replied:
thanks, fixed

patriknw reviewed (commented):
Just a few minor code comments so far. I don't have the mental capacity to understand this today. It was an intense week.

Can you describe at a somewhat high level what the tombstones are used for.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -34,11 +80,16 @@ object ORMap {
-  extends ReplicatedData with ReplicatedDataSerialization with RemovedNodePruning {
+  private[akka] val keys:          ORSet[A],
+  private[akka] val values:        Map[A, B],
+  private[akka] val tombstoneVals: Map[A, B],
```
@gosubpl I looked through this again and the part I don't fully understand is the tombstones. When are they removed (well, I see that they are removed at several places)?

gosubpl replied:
I will elaborate on tombstones in the summary comment below.

gosubpl commented:
@patriknw many thanks for taking a look at this PR. I have included suggested improvements. I have also added type tagging to the `zero` of `ORMap.DeltaOp`, because otherwise CRDTs built only from deltas would have different type, e.g. `ORMap[String, ORSet[String]]` instead of `ORMultiMap[String, String]`. This is tested in `ReplicatorMapDeltaSpec`, which is a copy of `ReplicatorDeltaSpec` but ported to `*Map` types.

As to the `tombstoneVals` - these are not 'classical' tombstones but rather a cache to use when merging with a `delta` update from another node. I have discovered the problem requiring them while porting the `PNCounterMapSpec` to deltas, when I was getting wrong final value of the `PNCounter` stored in the map. It can also be seen when using `GSet` or `ORSet` as a value type - please note the spec `not have anomalies for remove+updated scenario and deltas` in `ORMapSpec` (and subsequent 9 variations of it). For each `key` delete (which could be either local or coming up via full or delta update) I am storing the current `value` related to this `key`. Later, when there is `update delta` propagated and results in `key` re-appearance, the `update delta` is applied on the `value` stored in tombstone, so that final result is the same as if full update was propagated. However, if there is `delete delta` (or `put delta` - because `put` is effectively _delete_ with subsequent _update_) registered for that node, then it means that the value stored in tombstone has been superseded (on the particular node) and therefore is not merged with the coming `update delta`.
Thanks for asking those detailed questions, I have just discovered one more small issue related to the tombstones - which I have corrected.

gosubpl commented on akka-distributed-data/src/main/protobuf/ReplicatedDataMessages.proto:
```diff
@@ -72,7 +72,30 @@ message ORMap {
+  message Entry {
+    required ORMapDeltaOp operation = 1;
+    required ORSet underlying = 2;
+    required UniqueAddress node = 3;
```
For application of `delta` upon `tombstoneVals`

gosubpl commented on akka-distributed-data/src/main/protobuf/ReplicatedDataMessages.proto:
```diff
@@ -72,7 +72,30 @@ message ORMap {
+    required ORMapDeltaOp operation = 1;
+    required ORSet underlying = 2;
+    required UniqueAddress node = 3;
+    required sint32 zeroTag = 4;
```
To pass concrete type of the `*Map` - required to properly type the `zero` used for initial value by the delta `Replicator`

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/LWWMap.scala:
```diff
@@ -141,6 +143,14 @@ final class LWWMap[A, B] private[akka] (
+  override def resetDelta: LWWMap[A, B] =
+    new LWWMap(underlying.resetDelta)
+
+  override def delta: Option[D] = underlying.delta map { tagDelta(_)(LWWMapTag) }
```
Tag delta with correct detailed type.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -205,6 +207,14 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
+  override def resetDelta: ORMultiMap[A, B] =
+    new ORMultiMap(underlying.resetDelta)
+
+  override def delta: Option[D] = underlying.delta map { tagDelta(_)(ORMultiMapTag) }
```
Tag delta with correct detailed type.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/PNCounterMap.scala:
```diff
@@ -124,6 +127,14 @@ final class PNCounterMap[A] private[akka] (
+  override def resetDelta: PNCounterMap[A] =
+    new PNCounterMap(underlying.resetDelta)
+
+  override def delta: Option[D] = underlying.delta map { tagDelta(_)(PNCounterMapTag) }
```
Tag delta with correct detailed type.

gosubpl commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMapSpec.scala:
```diff
@@ -110,6 +141,280 @@ class ORMapSpec extends WordSpec with Matchers {
       merged2.entries("c").elements should be(Set("C"))
     }
 
+    "not have anomalies for remove+updated scenario and deltas" in {
```
10 scenarios requiring `tombstoneVals`

gosubpl commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMapSpec.scala:
```diff
@@ -110,6 +141,280 @@ class ORMapSpec extends WordSpec with Matchers {
+      val merged3 = m3 mergeDelta m4.delta.get
+
+      merged3.entries("a").elements should be(Set("A"))
+      merged3.entries("b").elements should be(Set("B2", "B3"))
```
If not for the value of `B2` saved in `tombstoneVals` I would get here `Set("B3")`

gosubpl replied:
Still works without tombstones, because now full value is being propagated, not its delta

gosubpl commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/PNCounterMapSpec.scala:
```diff
@@ -46,6 +46,24 @@ class PNCounterMapSpec extends WordSpec with Matchers {
+
+      // but if there is a conflicting update the entry is not removed
+      val m4 = merged1.resetDelta.increment(node2, "b", 10)
+      (m3 mergeDelta m4.delta.get).entries should be(Map("a" → 1, "b" → 13, "c" → 7))
```
Anomalous result of `"b" -> 10` without `tombstoneVals`

gosubpl replied:
Still works without tombstones, because now full value is being propagated, not its delta

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,77 @@ object ORMap {
+    override def zero: DeltaReplicatedData
+  }
+
+  private[akka] final val VanillaORMapTag = 0
```
Tags for tagging `zero` in deltas

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,77 @@ object ORMap {
+
+  /**
+   * INTERNAL API
+   */
```
Helper tag function

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -100,8 +179,18 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+      }
+      // put forcibly damages history, so we remove added values from the tombstone
+      // and propagate full value that will overwrite previous values
+      new ORMap(keys.add(node, key), values.updated(key, value), tombstoneVals - key, Some(newDelta))
```
Put forcibly damages history, let's not keep it in the tombstone, hence `tombstoneVals - key`

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -125,11 +214,42 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
-    val newValue = values.get(key) match {
-      case Some(old) ⇒ modify(old)
-      case _         ⇒ modify(initial)
+    val (oldValue, hasOldValue) = values.get(key) match {
```
`updated` with no previous value is a `put`. This will be indicated by `hasOldValue`

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -125,11 +214,42 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        modifiedDelta
+    }
+    // if updated is addition of previously removed key we need to remove the tombstone
+    new ORMap(keys.add(node, key), values.updated(key, newValue), tombstoneVals - key, Some(newDelta))
```
After `update` the value is now surely whole - we will prefer it to whatever comes from other node. So no need to keep previous state in tombstone. Hence `tombstoneVals - key`

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -150,7 +270,16 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        RemoveDeltaOp(keys.resetDelta.remove(node, key).delta.get, Map.empty[A, B], node)
+    }
+    // tombstone value for the key being removed
+    val newTombstoneVals = if (values.contains(key)) tombstoneVals + (key → values(key)) else tombstoneVals
```
We need to keep tombstone of removed value in order to merge it later with anything that comes in an update from the node that does not know about removal.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    // tombstone the values for removed keys
```
If anything was removed as a result of a merge - we tombstone it.

patriknw replied:
How long are such tombstone values kept? Looks like removal is triggered by other actions, so do I understand it correctly that if no more updates are performed the tombstone will live forever? I'm worried about growing full state size when different keys are added and removed.

Are there any assumptions that influence correctness with regards to the order or when the full state is delivered/merged and when the delta is delivered/merged. Note that several deltas, e.g. 17, 18, 19 can be grouped together and the full state corresponding to 18 can be applied first and then the delta group 17,18,19 is merged into that.

BTW, it's looking great, I just want to understand that it is rock solid.
I'll take the time to debug what happens in the tests, to get a better understanding.

gosubpl replied:
Unfortunately, tombstones are (as I have implemented them) a global cluster property - the only way to decide to delete the tombstone before applying `update` / `put` message for a given key would be to make sure the `delete` operation was applied on all cluster nodes - because then no operation coming from outside nodes will need tombstone (so wait for confirmation from all currently present cluster members for the reception of `delete` operation (either via delta, full or gossip)). This perhaps looks like a good job for `Replicator`. Let's discuss this avenue... I will also try to look into that (but then probably we need to get your delta confirmation feature merged before this).
Second thing - I will add some tests to validate that correctness assumptions are in no way violated in the scenario you are describing.

patriknw replied:
That would be complicated, since the cluster membership is dynamic. No, that is the wrong avenue. Could we get away without tombstones altogether if we don't use deltas for `put` and `remove`?

gosubpl replied:
Already `put` and `remove` propagate _delta_ for the `keys` `ORSet` but _full state_ for the `value`. Only `update` uses _delta-delta_ setup, and this is what results in anomaly, because we get only _delta_ of the second update but would lose the context.
I was thinking that perhaps it would be better to change existing semantics (and maybe a bit of an API). If every `DeltaReplicatedData` type had `clear` operation, it could be used to clear the value before removing the key. The relevant deltas would be propagated and we would be consistent in the operation semantics between full updates and deltas. However, this approach would have two drawbacks - for every key ever added we would keep forever at minimum its stub (consisting of key entry in the `keys` `ORSet` and the `vvector` of the `value`), and the semantics would be changed to "removal wins". So far the only way of getting rid of tombstones I figured out would be to not propagate the deltas of the `values` (but that wouldn't be a big win over not propagating deltas at all) or allow different behaviours for delta updates vs full updates (probably undesirable).

gosubpl replied:
I will have sometime tomorrow to dig through all _delta papers_ again to search for all information relevant to the `*Map` setup. If I correctly recall there was at least some guidance there.

patriknw replied:
Thanks for insights and investigation. I think propagation of full value of updated entry compared to propagation of the full state (full Map) is useful. Perhaps we should start simple and go with that for the 2.5.0 release, and later add some other Map with different semantics.

I think it's important that the Map can be used for dynamically adding/removing keys without growing the full state, since the top map of the Replicator doesn't have this dynamic aspect.

gosubpl replied:
I will remove the tombstones and revert the vanilla `ORMap` to `delta key` - `full value` setup. However, as `ORMultiMap` already does the _clear_ trick:
```
  @InternalApi private[akka] def put(node: UniqueAddress, key: A, value: Set[B]): ORMultiMap[A, B] = {
    val newUnderlying = underlying.updated(node, key, ORSet.empty[B]) { existing ⇒
      value.foldLeft(existing.clear(node)) { (s, element) ⇒ s.add(node, element) }
    }
    new ORMultiMap(newUnderlying)
  }
```
I will keep it on `delta` - `delta` scheme, which for that case will not require tombstones and will not impact semantics.

patriknw replied:
sounds great, thanks

gosubpl replied:
ok, so I have removed the tombstones, but I need to do some more tweaking & tests on `ORMultiMap` and remove `node` from ORMap's deltas. I will give heads-up when this PR is again ready for review.
Please note that the _clear trick_ for `ORMultiMap` didn't break any test and probably will not break any users' behaviour (would probably be good to ping Chris  and his team to double check when this is ready, as I expect them to be the main consumers of the Multi Map in the wild)

patriknw replied:
excellent

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    var thatValueDeltas: Map[A, List[(A, ReplicatedData)]] = Map.empty
+    var invalidatedTombstones: Map[A, Set[UniqueAddress]] = Map.empty
+
+    val processDelta: PartialFunction[ORMap.DeltaOp, Unit] = {
```
both `putOp` and `removeOp` will invalidate the tombstone for the node from which they are coming from. This means that further `updated` deltas from that node will not be merged with tombstoned state. This is because in that light cone the previous state got destroyed either by `put` or `remove`.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+          invalidatedTombstones += (key → Set(removeOp.node))
+      case updateOp: UpdateDeltaOp[A, _] ⇒
+        val key = updateOp.value.head._1
+        val value =
```
Merge with tombstone but not if it was invalidated for that node.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    new ORMap[A, B](thatKeys, mergedValues, tombstoneVals -- mergedKeys.elementsMap.keys.toIterator)
```
If we removed anything in the course of `mergeDelta`, we need to tombstone it.

gosubpl commented:
@patriknw I have added some comments in the review. If you wish them to be code comments instead, please let me know.

patriknw commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMapSpec.scala:
```diff
@@ -364,6 +364,56 @@ class ORMapSpec extends WordSpec with Matchers {
+      merged6.entries("a").elements should be(Set("A"))
+      // this time it's different...
+      merged6.entries("b").elements should be(Set("B2"))
+      merged6.entries("c").elements should be(Set("C"))
```
good illustration, I think that is "expected" - the simplified thinking is that you have to be prepared for that removals can be lost

gosubpl replied:
yes, but for the brand new `ORMultiMapWithDeltaDelta` this will be fixed... and will work consistent independent of timing

gosubpl commented:
@patriknw more ready for another round of review than ever before :)

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -115,10 +133,10 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
    */
   @InternalApi private[akka] def put(node: UniqueAddress, key: A, value: Set[B]): ORMultiMap[A, B] = {
-    val newUnderlying = underlying.updated(node, key, ORSet.empty[B]) { existing ⇒
+    val newUnderlying = underlying.updated(node, key, ORSet.empty[B], valueDeltas = true) { existing ⇒
```
I have just noticed this, and will fix it together with "small things" -> `valueDeltas` should depend on `mapTag == ORMultiMap.DeltaDelta... Tag`

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -156,8 +180,8 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
   @InternalApi private[akka] def addBinding(node: UniqueAddress, key: A, element: B): ORMultiMap[A, B] = {
-    val newUnderlying = underlying.updated(node, key, ORSet.empty[B])(_.add(node, element))
-    new ORMultiMap(newUnderlying)
+    val newUnderlying = underlying.updated(node, key, ORSet.empty[B], valueDeltas = true)(_.add(node, element))
```
fix needed

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -179,13 +203,17 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
   @InternalApi private[akka] def removeBinding(node: UniqueAddress, key: A, element: B): ORMultiMap[A, B] = {
     val newUnderlying = {
-      val u = underlying.updated(node, key, ORSet.empty[B])(_.remove(node, element))
+      val u = underlying.updated(node, key, ORSet.empty[B], valueDeltas = true)(_.remove(node, element))
```
fix needed

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -100,8 +186,18 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+        case None ⇒
+          putDeltaOp
+      }
+      // put forcibly damages history, so we remove added values from the tombstone
```
fix this comment, there are no more tombstones

gosubpl replied:
resolved

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -124,12 +220,43 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    // we can emit (and later merge) their deltas instead of full updates.
+    // However to avoid necessity of tombstones, the derived map type needs to support this
+    // with clearing the value (e.g. removing all elements if value is a set)
+    // before removing the key - like e.g. ORMultiMap does
```
`ORMultiMap.emptyWithDeltaDelta`

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def tagDelta(op: DeltaOp)(tag: Int) = op match {
```
is there a reason for second parameter list, otherwise make it two param

gosubpl replied:
Visual aesthetics ;) Wanted to have something like `map { tagDelta(deltaTag) _ }. Forgot that Scala does not do autocurrying like Haskell or ML. No need to fix, as deltas are now tagged at the source and this function is no more...

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+
+  // DeltaGroup is effectively a causally ordered list of individual deltas
+  /** INTERNAL API */
+  @InternalApi private[akka] final case class DeltaGroup[A, B <: ReplicatedData](ops: immutable.IndexedSeq[DeltaOp]) extends DeltaOp {
```
make `DeltaGroup` in `ORSet` internal

gosubpl replied:
sure

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+    }
+
+    override def merge(that: DeltaOp): DeltaOp = that match {
+      case other: AtomicDeltaOp[A, B] ⇒ DeltaGroup(Vector(this, other))
```
can't some be merged into one op, e.g. UpdateDeltaOp + UpdateDeltaOp?
see ORSet

patriknw replied:
that could especially be an advantage when updating same key frequently

gosubpl replied:
Would be happy to have this working, but I am not sure exactly how to implement this at the moment. Maybe a separate PR would be actually a good idea.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+  @InternalApi private[akka] final case class UpdateDeltaOp[A, X <: ReplicatedDelta](underlying: ORSet.DeltaOp, value: Map[A, X], zeroTag: Int = VanillaORMapTag) extends AtomicDeltaOp[A, X] {
+  }
+
+  // RemoveDeltaOp does not contain any value at all - the propagated 'value' map is empty
```
if so why is the Map needed?

gosubpl replied:
Not needed, removed :)

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -124,12 +220,43 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    def gatherDelta(newValue: B, newValueDelta: Option[ReplicatedDelta]) = if (newValueDelta.isDefined && hasOldValue)
+      (newValue, UpdateDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValueDelta.get)))
+    else
+      (newValue, PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue)))
```
looks like `match` would be useful

```scala
      val op = newValueDelta match {
        case Some(d) if hasOldValue => UpdateDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key →d))
        case _ => PutDeltaOp(keys.resetDelta.add(node, key).delta.get, Map(key → newValue))
      }
```

gosubpl replied:
resolved, thanks for the snippet

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -124,12 +220,43 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+
+    val (newValue, modifiedDelta) = if (valueDeltas && oldValue.isInstanceOf[DeltaReplicatedData]) {
+      val (newValue, newValueDelta) = modifyWithDelta(oldValue.asInstanceOf[DeltaReplicatedData])
+      gatherDelta(newValue, newValueDelta)
```
the `gatherDelta` reads nicely but to avoid the allocation of tuple instance I think it should be inlined (only used at one place), same with `modifyWithDelta`

gosubpl replied:
Those two functions were leftover from when there were more cases and the code would be unitelligible. Merged into main code flow now. Howerver, coming from C++ world I find such problems puzzling - why compiler cannot do this (simple) task...

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -150,13 +277,39 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def removeKey(node: UniqueAddress, key: A): ORMap[A, B] = {
```
add comment that this is only used for delta-values

gosubpl replied:
resolved

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,10 +327,107 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    dryMerge(that, mergedKeys, mergedKeys.elementsMap.keysIterator)
+  }
+
+  def mergeRetainingDeletedValues(that: ORMap[A, B]): ORMap[A, B] = {
```
private or protected

gosubpl replied:
resolved

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,10 +327,107 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    // helper function to simplify folds below
+    def foldValues(values: List[(A, ReplicatedData)], initial: B) =
+      values.foldLeft(initial) {
+        (acc: B, rhs: (A, ReplicatedData)) ⇒
```
```scala
      values.foldLeft(initial) {
        case (acc: DeltaReplicatedData, (_, value: ReplicatedDelta)) ⇒
          acc.mergeDelta(value.asInstanceOf[acc.D]).asInstanceOf[B]
        case (acc, (_, value)) ⇒
          acc.merge(value.asInstanceOf[acc.T]).asInstanceOf[B]
      }
```

gosubpl replied:
resolved, thanks for the snippet

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -186,10 +436,11 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
       case (_, data: RemovedNodePruning) ⇒ data.needPruningFrom(removedNode)
       case _                             ⇒ false
     }
+    keys.needPruningFrom(removedNode) || values.exists { dataNeedsPruning }
```
not sure this change is an improvement

gosubpl replied:
sorry, that was a leftover from the short period where I needed also to prune them tombstones. Reverted.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -3,22 +3,26 @@
 
-  val _empty: ORMultiMap[Any, Any] = new ORMultiMap(ORMap.empty)
+  val _empty: ORMultiMap[Any, Any] = new ORMultiMap(ORMap.empty, VanillaORMultiMapTag)
+  val _emptyWithDeltaDelta: ORMultiMap[Any, Any] = new ORMultiMap(ORMap.empty, DeltaDeltaORMultiMapTag)
```
`emptyWithDeltaValues` or `emptyWithValueDelta` might be a better name?

gosubpl replied:
Can this be `emptyWithValueDeltas` ? (plural of the second one :) )

patriknw replied:
sounds good

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -3,22 +3,26 @@
    * Provides an empty multimap.
    */
   def empty[A, B]: ORMultiMap[A, B] = _empty.asInstanceOf[ORMultiMap[A, B]]
+  def emptyWithDeltaDelta[A, B]: ORMultiMap[A, B] = _emptyWithDeltaDelta.asInstanceOf[ORMultiMap[A, B]]
```
document the difference compared to `empty` including the tradefoffs (growing state etc)

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -32,6 +36,9 @@ object ORMultiMap {
   }
+
+  private[akka] final val VanillaORMultiMapTag = 0
+  private[akka] final val DeltaDeltaORMultiMapTag = 1
```
any advantage of using integers here compared to a `Boolean`?

gosubpl replied:
actually, no :)

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -41,27 +48,35 @@ object ORMultiMap {
    */
   def entries: Map[A, Set[B]] =
-    underlying.entries.map { case (k, v) ⇒ k → v.elements }
+    underlying.entries.filter { case (k, _) ⇒ underlying.keys.elements.contains(k) }.map { case (k, v) ⇒ k → v.elements }
```
that's only needed if using delta-values, right?
if so, make it an `if`

use `collect` instead of `filter + map`

gosubpl replied:
`if` + `collect` :)

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -41,27 +48,35 @@ object ORMultiMap {
     import scala.collection.JavaConverters._
     val result = new java.util.HashMap[A, java.util.Set[B]]
-    underlying.entries.foreach {
+    underlying.entries.filter { case (k, _) ⇒ underlying.keys.elements.contains(k) }.foreach {
```
same here

gosubpl replied:
same here

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -71,7 +86,10 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
    */
   def get(key: A): Option[Set[B]] =
-    underlying.get(key).map(_.elements)
+    if (underlying.keys.elements.contains(key))
```
if on the delta-value aspect?

gosubpl replied:
yes, otherwise the other one - I have combined this in one if

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -80,11 +98,11 @@ final class ORMultiMap[A, B] private[akka] (private[akka] val underlying: ORMap[
     get(key).getOrElse(default)
 
-  def contains(key: A): Boolean = underlying.contains(key)
+  def contains(key: A): Boolean = underlying.keys.elements.contains(key)
```
should `contains` in `underlying` be changed instead?
is this only needed for delta-values or is it a good change in general?

gosubpl replied:
this is only needed for delta-values. However, this changed version will work universally. I wouldn't be eager to change this in the underlying, as people might need this divergence

patriknw reviewed (commented):
Here is a first round of comments. I'll go through serialization and tests in next round.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def tagDelta(op: DeltaOp)(tag: Int) = op match {
+    case p: PutDeltaOp[_, _]       ⇒ p.copy(zeroTag = tag)
```
I think we can avoid these matches, by adding
`def withZeroTag(tag: Int): DeltaOp` in `DeltaOp` and implement it in the concrete types, e.g. UpdateDeltaOp:

    override def withZeroTag(tag: Int): UpdateDeltaOp = copy(zeroTag = tag)

but I'm also thinking of why the tagging is done afterwards. Couldn't `ORMap` have a constructor parameter with the `zeroTag`, e.g. `PNCounterMap` knows it's tag when it creates the `underlying: ORMap`. Then `ORMap` could use the right tag immediately when creating the `DeltaOp` instances?

gosubpl replied:
I went the way of tagging the `underlying`. Works like a charm.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,85 @@ object ORMap {
+    def value: Map[A, B]
+    def zeroTag: Int
+    override def zero: DeltaReplicatedData = zeroTag match {
+      case PNCounterMapTag         ⇒ PNCounterMap.empty[A]
```
we could get rid of these matches by making `zeroTag` something richer than an `Int`, such as
```
trait ZeroTag {
  def zero: DeltaReplicatedData
  def value: Int
}

case object PNCounterMapTag extends ZeroTag {
  override def zero: DeltaReplicatedData = PNCounterMap.empty
  override def value: Int = 1
}
```

Another advantage of this is that it would be easier to open it up for custom implementations (outside akka) if needed.

The serialized format would still just be `Int` and match in the deserialization to pick the right instance.

gosubpl replied:
Thanks for the tip, implemented.

patriknw reviewed (commented):
Looked through serialization and tests

gosubpl commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializerSpec.scala:
```diff
@@ -184,6 +184,23 @@ class ReplicatedDataSerializerSpec extends TestKit(ActorSystem(
+        .resetDelta
+        .updated(address1, "a", ORSet.empty[String])(_.add(address1, "C")).delta.get)
+      // use Flag for this test as object key because it is serializable
+      checkSerialization(ORMap().put(address1, Flag(), GSet() + "A").delta.get)
```
something added

gosubpl commented on akka-distributed-data/src/test/scala/akka/cluster/ddata/ORMapSpec.scala:
```diff
@@ -110,6 +141,330 @@ class ORMapSpec extends WordSpec with Matchers {
+      val Flag(d3) = m3.entries("a")
+      d3 should be(true)
+    }
+
```
thanks

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -428,6 +397,13 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+      d.merge(deltaOp)
+    case None ⇒
+      deltaOp
+  }
```
nice

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -49,47 +46,50 @@ object ORMultiMap {
-      case (k, v) ⇒ result.put(k, v.elements.asJava)
-    }
+    if (withValueDeltas)
+      underlying.entries.collect { case (k, v) if underlying.keys.elements.contains(k) ⇒ result.put(k, v.elements.asJava) }
```
ah, this is actually not a `collect` since it's not returning anything

    underlying.entries.foreach { case (k, v) => if (underlying.keys.elements.contains(k)) result.put(k, v.elements.asJava) }

gosubpl replied:
indeed, thanks for spotting. Let's celebrate the side effects!

patriknw reviewed (commented):
changes look good so far, just a small scala style thingy

gosubpl commented on akka-distributed-data/src/main/protobuf/ReplicatedDataMessages.proto:
```diff
@@ -72,7 +72,29 @@ message ORMap {
+    optional OtherMessage value = 5;
+    optional sint32 intKey = 6;
+    optional sint64 longKey = 7;
+    optional OtherMessage otherKey = 8;
```
That is not exactly true, `ORMultiMap.Entry` has also an `ORSet` which is not needed here. In addition to that - we are now having up to one `entryData` but later, for combining `updateDeltaOp` s we will need to have proper map here, but still only one `ORSet`, so the `ORMultiMap.Entry` cannot be re-used. But I have re-factored the common part into `ORMapDeltaGroup.MapEntry` which is now `optional` but later can be made `repeated`.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -174,10 +327,107 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
+    val processDelta: PartialFunction[ORMap.DeltaOp, Unit] = {
+      case putOp: PutDeltaOp[A, B] ⇒
+        val key = putOp.value.head._1
+        val value = putOp.value.head
```
Not sure about merging (see comment for `updateDeltaOp`. However, for the time being I have fixed that so that `PutDeltaOp` has just a pair and not a map inside.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/protobuf/ReplicatedDataSerializer.scala:
```diff
@@ -536,9 +573,131 @@ class ReplicatedDataSerializer(val system: ExtendedActorSystem)
+    else if (eh.hasLongKey(entry)) Some(eh.getLongKey(entry) → valueCreator(eh.getValue(entry)))
+    else if (eh.hasOtherKey(entry)) Some(otherMessageFromProto(eh.getOtherKey(entry)) → valueCreator(eh.getValue(entry)))
+    else None
+    if (elem.isDefined)
```
Sure. I am wondering why consistently go from `if` to `match` ? Any particular reasons beyond aesthetics / consistent style ? Are there any problems for Scala with `if` that `match` doesn't have?

patriknw replied:
`Option.get` is not idiomatic Scala (it's a bad smell, because it must almost be accompanied with `.isDefined`.

Most idiomatic is `.getOrElse`, but we avoid that because it's allocating a function instance (yeah, we are at an abnormal level with regards to performance concerns). In Scala 2.12 I think that is not an issue any more.

Therefore we prefer `match` on `Option`.

gosubpl replied:
Thanks for the explanation. I would also rather not allocate if I can avoid it, but I am surprised that this has ever been a problem in Scala (i.e. compiler not optimizing this out) - good to know that both language and tools are maturing :)

gosubpl commented:
@patriknw things that are missing are aggregation / coalescing of deltas, some comments (especially on the tradeoffs for `ORMultiMap.withValueDeltas`) and maybe some docs. If you would like to merge this soon, please do that and I will create a follow-up PR(s) with all of these (comments/docs/coalescing).
Depending on how much time I have later today, I also might be able to add some comments/docs while this PR is still being reviewed and follow-up only with the delta coalescing stuff.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,125 @@ object ORMap {
+
+  // PutDeltaOp contains ORSet delta and full value
+  /** INTERNAL API */
+  @InternalApi private[akka] final case class PutDeltaOp[A, B <: ReplicatedData](underlying: ORSet.DeltaOp, value: (A, B), zeroTag: ZeroTag = VanillaORMapTag) extends AtomicDeltaOp[A, B] {
```
I think these should not have default param value. Easier to make sure that the right tag is used if it's an required param

gosubpl replied:
Agreed. When I was introducing those, I have also omitted the tag, exactly for the reason you are giving. Will fix.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -34,11 +156,16 @@ object ORMap {
-  extends ReplicatedData with ReplicatedDataSerialization with RemovedNodePruning {
+  private[akka] val keys:    ORSet[A],
+  private[akka] val values:  Map[A, B],
+  private[akka] val zeroTag: ZeroTag               = ORMap.VanillaORMapTag,
```
same here, no default param value

gosubpl replied:
Same as above.

patriknw commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMultiMap.scala:
```diff
@@ -3,22 +3,26 @@
    * Java API
    */
   def create[A, B](): ORMultiMap[A, B] = empty[A, B]
+  def createWithDeltaDelta[A, B](): ORMultiMap[A, B] = emptyWithValueDeltas[A, B]
```
should be `createWithValueDeltas`, but I think `emptyWithValueDeltas` can be used for Java also, so it can be removed

gosubpl replied:
Agreed. I will try to add some Java tests in the follow-up to be on the safe side with the Java API :)

patriknw reviewed (approved):
LGTM, I reviewed everything one more time.
Awesome contribution @gosubpl. Great thinking and lot's of hard work.

I'll fix the minor Java API issue.

You can follow up with another PR with the remaining minor things and see if it's possible to optimize the delta coalescing, but nothing of that is urgent for the release. Have a nice weekend and relax.

patriknw commented:
I'm running this branch in a jenkins repeat job, since it wasn't included in the nightly builds. Looking good. https://jenkins.akka.io:8498/job/akka-multi-node-repeat/

gosubpl commented:
@patriknw many thanks for reviewing so thoroughly. I learned a lot improving my code! I must confess that I have enjoyed greatly working on getting _deltas_ into Akka's CRDTs :) I will submit follow-up PR(s) as soon as I have a spare moment again :)

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,125 @@ object ORMap {
+  @InternalApi private[akka] case object LWWMapTag extends ZeroTag {
+    override def zero: DeltaReplicatedData = LWWMap.empty
+    override final val value: Int = 4
+  }
```
Yes, moving them to target types will make it easier to offer a way of creating user-defined concrete map types.

gosubpl commented on akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala:
```diff
@@ -22,6 +25,125 @@ object ORMap {
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def emptyWithLWWMapTag[A, B <: ReplicatedData]: ORMap[A, B] = new ORMap(ORSet.empty, Map.empty, zeroTag = LWWMapTag)
```
Thanks, will fix.

gosubpl commented on akka-distributed-data/src/main/protobuf/ReplicatedDataMessages.proto:
```diff
@@ -72,47 +72,73 @@ message ORMap {
   }
   
```
I have just added a grouping into `MapEntry`. This will allow to extend easier with coalesced deltas.
